linux/src/completion_port_linux.c

```diff
@@ -0,0 +1,434 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+#include <stddef.h>
+#include <inttypes.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <sys/epoll.h>
+#include <sys/socket.h>
+
+#ifdef USE_VALGRIND
+#include "valgrind/helgrind.h"
+#endif
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/containing_record.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/refcount.h"
+#include "c_pal/socket_handle.h"
+#include "c_pal/s_list.h"
+#include "c_pal/sync.h"
+#include "c_pal/threadapi.h"
+
+#include "c_pal/completion_port_linux.h"
+
+#define MAX_EVENTS_NUM      64
+#define EVENTS_TIMEOUT_MS   2*1000
+
+typedef struct EPOLL_THREAD_DATA_TAG
+{
+    S_LIST_ENTRY link;
+    ON_COMPLETION_PORT_EVENT_COMPLETE event_callback;
+    void* event_callback_ctx;
+    SOCKET_HANDLE socket;
+} EPOLL_THREAD_DATA;
+
+typedef struct COMPLETION_PORT_TAG
+{
+    int epoll;
+    THREAD_HANDLE thread_handle;
+    volatile_atomic int32_t worker_thread_continue;
+    S_LIST_ENTRY alloc_data_list;
+    volatile_atomic int32_t recv_data_access;
+    volatile_atomic int32_t pending_calls;
+} COMPLETION_PORT;
+
+MU_DEFINE_ENUM_STRINGS(COMPLETION_PORT_EPOLL_ACTION, COMPLETION_PORT_EPOLL_ACTION_VALUES)
+
+DEFINE_REFCOUNT_TYPE(COMPLETION_PORT);
+
+static int add_thread_data_to_list(COMPLETION_PORT* completion_port, EPOLL_THREAD_DATA* epoll_data)
+{
+    int result;
+    do
+    {
+        int32_t current_val = interlocked_compare_exchange(&completion_port->recv_data_access, 1, 0);
+        if (current_val == 0)
+        {
+            break;
+        }
+        else
+        {
+            // Do Nothing wait for address
+        }
+        (void)wait_on_address(&completion_port->recv_data_access, current_val, UINT32_MAX);
+    } while (true);
```

Comment:
This kind of looks like some sort of "acquire lock" or "enter critical section" function, and at the end those 2 lines seem like a "release lock" or "leave critical section".

At the very least I'd consider making a pair of static functions named as such (if not a full blown module), especially since it's repeated below.